pkg/meta/meta.go

```diff
@@ -63,6 +68,18 @@ func ReferenceTo(o metav1.Object, of schema.GroupVersionKind) *corev1.ObjectRefe
 	}
 }
 
+// TypedReferenceTo returns a typed object reference to the supplied object,
+// presumed to be of the supplied group, version, and kind.
+func TypedReferenceTo(o metav1.Object, of schema.GroupVersionKind) *v1alpha1.TypedReference {
+	v, k := of.ToAPIVersionAndKind()
+	return &v1alpha1.TypedReference{
+		APIVersion: v,
+		Kind:       k,
+		Name:       o.GetName(),
```

Comment:
```suggestion
		Name:       o.GetName(),
		Namespace: o.GetNamespace(),
```

My immediate expectation from a `TypedReferenceTo` function is for it to have a full reference but it doesn't include namespace. So, I think namespace should be included unless there is an explicit sign(in its name or how argument list is constructed) that says this function is specifically for cluster-scoped resources. Also, it cannot be a full replacement of `ReferenceTo` if it doesn't include namespace.

Reply:
It sounds like you have issue with the `TypedReference` type rather than this function. `TypedReference` does not have a namespace so one cannot be returned here. In regards to replacing `ObjectReference`, I don't think we are saying `TypedReference` should be used in all cases instead of `ObjectReference`. We are saying that a more scoped reference should be used instead on `ObjectReference`, so if one does not exist that fits your use case then you should add it.

Reply:
I'm actually fine with `TypedReference` being cluster-scoped but I don't think it replaces `ObjectReference` in this specific context. For example, if I used `ReferenceTo` for connection secrets then this deprecation means that I have to implement my own function now instead of using the suggested `TypedReferenceTo` function. So, I'd be fine with this function if it didn't aim to deprecate & replace `ReferenceTo`.

Reply:
@muvaf the deprecation notice does not indicate that you should always use `TypedReference` or `TypedReferenceTo`. It gives it as an example, but says to just use a generally more scoped reference type. We _are_ saying that if you want a different reference type then you should add it and a helper function for it. We are trying to moved away from using `ObjectReference` per the issue linked by @negz

Reply:
Alright, I think we need the namespaced alternative but that doesn't seem to be in the scope of this PR. I guess we can open another one that adds the necessary type and helper function when needed by core Crossplane.

Reply:
I am happy to add it in this PR if we have a use case for it.

Reply:
I think all logics related to connection secret creation rely on that function for getting an owner ref of claim, like [here](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/resource/resource.go#L119). Also, I use it in crossplane agent, too, for setting owner ref of the propagated secret to the claim.

Reply:
Hmm for the sake of ownerReferences, it may make sense to no deprecate this, but rather just say "please use a more scoped reference if possible. What do you think? /cc @negz

Reply:
Well, I just realized that `OwnerReference` struct doesn't have `Namespace` field anyway since it's either namespace<>namespace or cluster-scoped<>cluster-scoped. So, replacing `AsOwner` and `AsController` functions with the ones that accept `TypedReference` could be sufficient to cover that use case. As you pointed out, if another thing comes up that requires namespace, we can consider an addition.

Reply:
@muvaf nice find! This goes a long way towards addressing https://github.com/crossplane/crossplane-runtime/issues/49 as well. Update coming!